### DISCUSSION TITLE
NestJS dependency issue with nest-commander

### DISCUSSION THREAD

**Discussion by KiriLucas** [jmcdo29/nest-commander#298] [Q&A]
Someone on stackoverflow advised me to ask this here, so, here we go:

I have five modules:

- `AppModule`
- `FirstModule`
- `SecondModule`
- `SourceModule1`
- `SourceModule2`

And one extra for `nest-commander`:

- `CommanderModule`, which has the following files:

- Job.command.ts (with the @Command() decorator and implements CommandRunner interface)
- CommanderService (Just a service layer that `Job.command` would call because of some other concerns)
- Commander.ts (With my bootstrap method, which is the one I call using my command)

And to call my command, I'm using: `npx ts-node src/<path> argument --option abc`

All modules have a service, and every module is imported on `AppModule`.
`FirstModule` is importing `SourceModule1` and `SourceModule2` and using them on it's service, the same goes for `SecondModule`, they both use `SourceModule` 1 and 2 to meet their needs.

My `CommanderModule` needs `FirstModule` because of `CommanderService`... When a specific command is ran, `FirstModule's service` shall be called by `CommandModule's service`.

The thing is, whenever I try using `FirstModule` on my `CommanderService` so `Job.command` may call a `CommanderService` method and that method may call `FirstModule's service`, my command doesn't reach it's "run" method, if I comment out `FirstModule's service`, it works.
I know `nest-commander` uses `@Command()` as a way of dealing with, among some other things, dependency injection, but shouldn't I be able to use a `CommanderService` besides the `Job.command?`

**Comment by jmcdo29** [comment]
Do you get an error you can show? Can you provide access to your code?

**Reply by KiriLucas** [reply]
No error at all, about my code:

```typescript
@Module({
  imports: [FirstModule],
  providers: [CommanderService, JobCommands],
})
export class CommanderModule {}

```

**AppModule:**
```typescript
@Module({
  imports: [
    FirstModule,
    SecondModule,
    SourceModule1,
    SourceModule2,
    CommanderModule
  ],
  providers: [],
})
export class AppModule {}
````

**SourceModule 1 and 2 looks almost the same:**

```typescript
@Module({
  imports: [],
  providers: [SourceProvider1],
  exports: [SourceProvider1],
})
export class SourceModule1 {}
````

**First and second module looks almost the same:**
```typescript
@Module({
  imports: [SourceModule1],
  providers: [FirstService],
  exports: [FirstService],
})
export class FirstModule {}
```

**Commander.ts:**
```typescript
(async function bootstrap() {
  await CommandFactory.run(CommanderModule, new Logger);
})();
```

**CommanderService:**
```typescript
export class CommanderService {
    constructor(
        private service: FirstService,
        ) { }

    async testificate(){
        console.log('Reached')
        console.log('COMMANDER SERVICE')
        this.service.placeHolder()
    }
}
```

Job.commands looks exactly the same as that on nest-commander docs, but theres one single line calling the CommanderService

**Reply by jmcdo29** [reply]
How does `CommanderService` get called? Does it have any decorators? Anything else about it?

**Reply by KiriLucas** [reply]
CommanderService gets called from `Job.command.ts`, just like this:

```typescript
import { Command, CommandRunner, Option } from 'nest-commander';
import { CommanderService } from './commander.service';

@Command({ name: 'callService', arguments: '<jobType>' })
export class JobCommands implements CommandRunner {
  constructor(private service: CommanderService) {}

  async run(inputs: string[], options: Record<string, any>): Promise<void> {
    this.service.testificate(); // Here is where I call CommanderService
}

  @Option({
    flags: '-target, --target <target>'
  })
  parseShell(val: string) {
    return val;
  }
}

```

**Reply by jmcdo29** [reply]
Does `CommanderService` have any decorators on it? Or do you just end up with `cannot call "placeHolder" of "undefined"` or something similar?

**Reply by micalevisk** [reply]
@KiriLucas btw if you want to inject `FirstService` into `CommanderService` you'll need to annotate `CommanderService` with `@Injectable()` (due to how _nestjs_ works).

**Reply by KiriLucas** [reply]
@jmcdo29 (and @micalevisk ) The only decorator it has is @Injectable(), sorry for cutting it out of my snippet! 

```typescript
@Injectable()
export class CommanderService {
    constructor(
        private service: FirstService,
        ) { }

    async testificate(){
        console.log('Reached')
        console.log('COMMANDER SERVICE')
        this.service.placeHolder()
    }
}
```

And when I run my command with npx, it goes as far as into the bootstrap method on `commander.ts`.

@jmcdo29 It won't call placeHolder method since it doesn't even reach the CommanderService 😞

**Reply by jmcdo29** [reply]
You should try `await this.service.testificate()`. If that's an async method that isn't being awaited, the node process will close out before `nextTick()` gets called and the async function will never actually get fired

**Reply by jmcdo29** [reply]
If that's not the fix, I'm gonna probably need a git repo I can work with

**Reply by KiriLucas** [reply]
Sorry for taking so long to reply, I found a possible reason for my issue:
First cause was because of my custom CommanderModule, as soon as I deleted CommanderModule and started using it as a AppModule provider, it worked. Not entirely sure why tough, I'll try to investigate this a bit.

Second and possibly not the actual cause if because of [node-rfc](https://github.com/SAP/node-rfc)... It has some compatibility issues with a few Nest libs, but that's probably no the case with `nest-commander` as it's now working.

Nevertheless, if you wish, I can create a repo with my code (I'm currently using a private Azure Devops repo, so I would need to take a look into the whole project and see what I'm allowed to share), just let me know!

**Reply by jmcdo29** [reply]
A reproduction of the non-function code would be wonderful if you can provide it
